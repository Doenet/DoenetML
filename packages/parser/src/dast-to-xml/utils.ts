import { stringifyEntitiesLight } from "stringify-entities";
import { DastNodes } from "../types";

/**
 * Escape a string.
 */
export function escape(
    value: string,
    subset: string[],
    unsafe?: RegExp | null | undefined,
): string {
    const result = clean(value);

    return unsafe ? result.replace(unsafe, encode) : encode(result);

    /**
     * Actually escape characters.
     */
    function encode(value: string): string {
        return (
            stringifyEntitiesLight(value, { subset })
                // We want fancy named versions of these two escaped characters
                .replace(/&#x3C;/g, "&lt;")
                .replace(/&#x26;/g, "&amp;")
        );
    }
}

const nonCharacter = /[\u0000-\u0008\u000B\u000C\u000E-\u001F]/g;
/**
 * Remove non-characters.
 */
export function clean(value: string) {
    return String(value || "").replace(nonCharacter, "");
}

/**
 * Encode a node name.
 */
export function name(value: string) {
    const subset = ["\t", "\n", " ", '"', "&", "'", "/", "<", "=", ">"];
    return escape(value, subset);
}

/**
 * Merge adjacent text nodes in an array
 */
export function mergeAdjacentTextInArray(nodes: DastNodes[]): DastNodes[] {
    const needsMerging = nodes.some(
        (n, i) => n.type === "text" && nodes[i + 1]?.type === "text",
    );
    if (!needsMerging) {
        return nodes;
    }
    const ret: DastNodes[] = [];
    for (let i = 0; i < nodes.length; i++) {
        let node = nodes[i];
        let nextNode = nodes[i + 1];
        if (!nextNode) {
            ret.push(node);
            continue;
        }
        if (node.type === "text" && nextNode.type === "text") {
            node = { ...node };
            ret.push(node);
            while (nextNode?.type === "text") {
                node.value += nextNode.value;
                if (node.position && nextNode.position) {
                    node.position.end = nextNode.position.end;
                }
                i++;
                nextNode = nodes[i + 1];
            }
        } else {
            ret.push(node);
        }
    }

    return ret;
}

/**
 * Recursively remove position info from a DAST tree.
 *
 * **Note**: this mutates the tree.
 */
export function filterPositionInfo(
    nodes: DastNodes | DastNodes[],
): DastNodes | DastNodes[] {
    if (typeof nodes !== "object" || nodes === null) {
        return nodes;
    }
    if (Array.isArray(nodes)) {
        return nodes.flatMap(filterPositionInfo);
    }
    if (nodes && typeof nodes === "object" && "position" in nodes) {
        delete nodes.position;
    }
    for (const value of Object.values(nodes)) {
        filterPositionInfo(value);
    }
    return nodes;
}
