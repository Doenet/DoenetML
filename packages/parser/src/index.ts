export * from "./parser";
export * from "./extract-dast-errors";
