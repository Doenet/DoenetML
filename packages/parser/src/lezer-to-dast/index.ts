export { lezerToDast } from "./lezer-to-dast";
export { lezerToDastV6 } from "./lezer-to-dast-v6";
