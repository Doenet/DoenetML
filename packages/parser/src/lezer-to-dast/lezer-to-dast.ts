import { SyntaxNode, TreeBuffer } from "@lezer/common";
import {
    CloseTag,
    MissingCloseTag,
    OpenTag,
    TagName,
} from "../generated-assets/lezer-doenet.terms";
import { parser } from "../generated-assets/lezer-doenet";
import {
    DastAttribute,
    DastDoctype,
    DastElementContent,
    DastFunctionMacro,
    DastMacro,
    DastNodes,
    DastRoot,
    DastRootContent,
    DastText,
    LezerSyntaxNodeName,
} from "../types";
import { createErrorNode } from "./create-error-node";
import {
    attributeValueText,
    createOffsetToPositionMap,
    entityToString,
    extractContent,
    extractDoctypeInfo,
    findFirstErrorInChild,
    getLezerChildren,
    lezerNodeToPosition,
    textNodeToText,
    updateNodePositionData,
} from "./lezer-to-dast-utils";
import { parseMacros } from "../macros";
import { gobbleFunctionArguments } from "./gobble-function-arguments";

/**
 * Create a lezer `SyntaxNode` from a string. This can be passed
 * to `lezerToDast` to create a DAST tree.
 */
export function stringToLezer(source: string): SyntaxNode {
    const tree = parser.parse(source);
    return tree.topNode;
}

/**
 * Convert a lezer `SyntaxNode` into a DAST tree.
 */
export function lezerToDast(
    node: SyntaxNode | string,
    source?: string,
): DastRoot {
    if (typeof node === "string") {
        const tree = parser.parse(node);
        source = node;
        return _lezerToDast(tree.topNode, source);
    }
    if (source == null) {
        throw new Error(
            `If you provide a SyntaxNode, you must also provide the source string`,
        );
    }
    return _lezerToDast(node, source);
}

function _lezerToDast(node: SyntaxNode, source: string): DastRoot {
    const offsetMap = createOffsetToPositionMap(source);
    return {
        type: "root",
        children: gobbleFunctionArguments(lezerNodeToDastNode(node)),
        position: lezerNodeToPosition(node, offsetMap),
    };

    function lezerNodeToDastNode(node: SyntaxNode): DastRootContent[] {
        if (!node) {
            throw new Error(`Expecting node but got ${JSON.stringify(node)}`);
        }
        const name = node.type.name as LezerSyntaxNodeName;
        switch (name) {
            case "Document":
                return getLezerChildren(node).flatMap(lezerNodeToDastNode);
            case "Element": {
                const openTag =
                    node.getChild("OpenTag") || node.getChild("SelfClosingTag");
                if (!openTag) {
                    console.warn(
                        `Could not find open tag for element ${extractContent(
                            node,
                            source,
                        )}`,
                    );
                    return [];
                }
                let children: DastElementContent[] = [];
                // The open tag may have an error in it.
                const openTagError = findFirstErrorInChild(openTag);
                if (openTagError) {
                    const errorNode = createErrorNode(
                        openTagError,
                        source,
                        offsetMap,
                    );
                    children.push(errorNode);
                }
                // If we have an open tag but no closing tag, it's an error.
                // It's not an error if the tag was self closing.
                let elementError =
                    node.getChild("âš ") || node.getChild(MissingCloseTag);
                if (
                    openTag.type.name === "OpenTag" &&
                    !node.getChild("CloseTag") &&
                    !openTagError &&
                    elementError
                ) {
                    const errorNode = createErrorNode(
                        elementError,
                        source,
                        offsetMap,
                    );
                    children.push(errorNode);
                }

                const tag = openTag.getChild("TagName");
                const name = tag ? extractContent(tag, source) : "";
                const attributes: DastAttribute[] = [];
                for (const attrTag of openTag.getChildren("Attribute")) {
                    const error = findFirstErrorInChild(attrTag);
                    if (error) {
                        const errorNode = createErrorNode(
                            error,
                            source,
                            offsetMap,
                        );
                        children.push(errorNode);
                    }
                    const attrName = attrTag.getChild("AttributeName");
                    const attrValue = attrTag.getChild("AttributeValue");
                    if (!attrName) {
                        continue;
                    }
                    let attrChildren: DastAttribute["children"] = attrValue
                        ? [
                              ...reprocessTextForMacros({
                                  type: "text",
                                  value: attributeValueText(attrValue, source),
                                  position: lezerNodeToPosition(
                                      attrValue,
                                      offsetMap,
                                  ),
                              } as DastText),
                          ]
                        : [];
                    // Attributes with no specified value are assigned the value "true".
                    // E.g. `<foo bar />` is the same as `<foo bar="true" />`
                    attributes.push({
                        type: "attribute",
                        name: extractContent(attrName, source),
                        children: attrChildren,
                        position: lezerNodeToPosition(attrTag, offsetMap),
                    });
                }
                // Children get pushed after attributes so that any attribute errors will
                // appear first.
                children.push(
                    ...getLezerChildren(node).flatMap(
                        (n) => lezerNodeToDastNode(n) as DastElementContent[],
                    ),
                );
                children = gobbleFunctionArguments(
                    children,
                ) as DastElementContent[];

                return [
                    {
                        type: "element",
                        name,
                        attributes,
                        children,
                        position: lezerNodeToPosition(node, offsetMap),
                    },
                ];
            }
            case "Text": {
                const textNode: DastText = {
                    type: "text",
                    value: textNodeToText(node, source),
                    position: lezerNodeToPosition(node, offsetMap),
                };
                return reprocessTextForMacros(textNode);
            }
            case "Ampersand":
                return [
                    {
                        type: "text",
                        value: "&",
                        position: lezerNodeToPosition(node, offsetMap),
                    },
                ];
            case "CharacterReference":
            case "EntityReference":
                return [
                    {
                        type: "text",
                        value: entityToString(node, source),
                        position: lezerNodeToPosition(node, offsetMap),
                    },
                ];
            case "ProcessingInst": {
                const fullContent = extractContent(node, source);
                let value = fullContent.slice(2, fullContent.length - 2);
                const match = value.match(/^[\w-]*/);
                const name = match?.[0] || "";
                if (name) {
                    value = value.slice(name.length).trim();
                }
                return [
                    {
                        type: "instruction",
                        name,
                        value,
                        position: lezerNodeToPosition(node, offsetMap),
                    },
                ];
            }
            case "Comment": {
                const fullContent = extractContent(node, source);
                return [
                    {
                        type: "comment",
                        value: fullContent.slice(4, fullContent.length - 3),
                        position: lezerNodeToPosition(node, offsetMap),
                    },
                ];
            }
            case "Cdata": {
                const fullContent = extractContent(node, source);
                return [
                    {
                        type: "cdata",
                        value: fullContent.slice(9, fullContent.length - 3),
                        position: lezerNodeToPosition(node, offsetMap),
                    },
                ];
            }
            case "DoctypeDecl": {
                // DocTypes look like <!DOCTYPE HTML PUBLIC '-//W3C//DTD HTML 4.0 Transitional//EN' 'http://www.w3.org/TR/REC-html40/loose.dtd'>
                const fullContent = extractContent(node, source);
                let value = fullContent.slice(10, fullContent.length - 1);
                const doctypeInfo = extractDoctypeInfo(value);
                return [
                    {
                        type: "doctype",
                        ...doctypeInfo,
                        position: lezerNodeToPosition(node, offsetMap),
                    } as DastDoctype,
                ];
            }
            case "Attribute":
            case "AttributeName":
            case "AttributeValue":
            case "EndTag":
            case "Is":
            case "OpenTag":
            case "SelfCloseEndTag":
            case "SelfClosingTag":
            case "TagName":
            case "StartTag":
            case "StartCloseTag":
                return [];
            case "CloseTag": {
                const errorNode = findFirstErrorInChild(node);
                if (errorNode) {
                    return [createErrorNode(errorNode, source, offsetMap)];
                }
                return [];
            }
            case "MismatchedCloseTag": {
                // We may have a mismatched tag, but the tag may also be incomplete itself.
                // For example `</x ` (with no closing `>`). In this case, we want the error
                // in the tag itself to take priority.
                const closeTagError = findFirstErrorInChild(node);
                if (closeTagError) {
                    return [createErrorNode(closeTagError, source, offsetMap)];
                }

                const parent = node.parent;
                const openTag = parent?.getChild(OpenTag);
                const closeTag = parent?.getChild(CloseTag);
                if (!parent || !openTag) {
                    const message = `Invalid DoenetML: Found closing tag \`${extractContent(
                        node,
                        source,
                    )}\`, but no corresponding opening tag`;
                    return [
                        {
                            type: "error",
                            message,
                            position: lezerNodeToPosition(node, offsetMap),
                        },
                    ];
                }
                // If we have a parent, check to see if we also have a close tag.
                // This could arise in code like `<foo></bar></foo>`
                if (closeTag) {
                    const message = `Invalid DoenetML: Found closing tag \`${extractContent(
                        node,
                        source,
                    )}\`, but no corresponding opening tag`;

                    return [
                        {
                            type: "error",
                            message,
                            position: lezerNodeToPosition(node, offsetMap),
                        },
                    ];
                }
                // In this case, there was an open tag, a mismatched close tag, but no actual close tag.
                // E.g. `<foo>bar</baz>`
                const tagNameTag = openTag.getChild(TagName);
                const openTagName = tagNameTag
                    ? extractContent(tagNameTag, source)
                    : "";
                const message = `Invalid DoenetML: Mismatched closing tag. Expected \`</${openTagName}>\`. Found \`${extractContent(
                    node,
                    source,
                )}\``;

                return [
                    {
                        type: "error",
                        message,
                        position: lezerNodeToPosition(node, offsetMap),
                    },
                ];
            }
            case "MissingCloseTag":
            case "âš ":
                return [];
            default:
                const unhandledName: never = name;
                console.log(
                    `Encountered Lezer node of unknown type ${unhandledName}`,
                );
        }
        return [];
    }
}

/**
 * Process a text node and expose any macros within as DastNodes.
 */
function reprocessTextForMacros(
    textNode: DastText,
): (DastText | DastMacro | DastFunctionMacro)[] {
    if (!textNode.value.includes("$")) {
        return [textNode];
    }
    // If there is a `$` in the text, it may contain macros, so re-parse it
    // looking for macros.
    const parsed = parseMacros(textNode.value);
    // The text node may be located anywhere in the source and we have
    // just re-parsed it, so we need to make sure the position data is
    // correct.
    parsed.forEach((n) => updateNodePositionData(n, textNode));
    return parsed;
}
