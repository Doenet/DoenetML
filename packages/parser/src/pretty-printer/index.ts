import { Options } from "prettier";
import * as Prettier from "prettier/standalone";
import { PrintOptions } from "../types";
import * as prettierPluginDoenet from "./prettier-plugin-doenet";

export async function prettyPrint(
    source = "",
    options: Options & PrintOptions = {},
) {
    const plugins = options.plugins || [];
    plugins.push(prettierPluginDoenet);

    return await Prettier.format(source, {
        printWidth: 80,
        useTabs: false,
        tabWidth: 4,
        ...options,
        parser: "doenet-parser",
        plugins,
    });
}
