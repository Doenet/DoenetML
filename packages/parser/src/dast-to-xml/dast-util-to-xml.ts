// Code modified from xast-util-to-xml MIT License https://github.com/syntax-tree/xast-util-to-xml
import { ccount } from "ccount";
import {
    DastAttribute,
    DastElement,
    DastNodes,
    DastMacroFullPath,
    DastFunctionMacro,
    DastMacro,
    DastMacroPathPart,
    PrintOptions,
    DastNodesV6,
    DastMacroV6,
    DastFunctionMacroV6,
} from "../types";
import { escape, mergeAdjacentTextInArray, name } from "./utils";
import { macroToString as macroToStringV6 } from "../macros-v6/macro-to-string";

/**
 * Serialize a xast tree to XML.
 */
export function toXml(
    tree?: DastNodes[] | DastNodes | DastNodesV6[] | DastNodesV6 | null,
    options?: PrintOptions,
) {
    options = options || {};
    if (!tree) {
        return "";
    }
    const node = (
        Array.isArray(tree) ? { type: "root", children: tree } : tree
    ) as DastNodes;

    return nodesToXml(node, options);
}

/**
 * Serialize a node.
 */
export function nodesToXml(
    node: DastNodes | DastNodes[],
    options: PrintOptions,
): string {
    if (Array.isArray(node)) {
        return mergeAdjacentTextInArray(node)
            .map((child) => nodesToXml(child, options))
            .join("");
    }

    const type = node && node.type;
    switch (node?.type) {
        case "root":
            return nodesToXml(node.children, options);
        case "cdata": {
            const unsafe = /]]>/g;
            const subset = [">"];
            return "<![CDATA[" + escape(node.value, subset, unsafe) + "]]>";
        }
        case "comment":
            return "<!--" + escape(node.value, ["-"]) + "-->";
        case "doctype": {
            const nodeName = name(node.name);
            const pub = node.public;
            const sys = node.system;
            let result = "<!DOCTYPE";

            if (nodeName !== "") {
                result += " " + nodeName;
            }

            if (pub) {
                result += " PUBLIC " + quote(pub);
            } else if (sys) {
                result += " SYSTEM";
            }

            if (sys) {
                result += " " + quote(sys);
            }

            return result + ">";
        }
        case "element": {
            const nodeName = name(node.name);
            const content = nodesToXml(node.children, options);
            const attributes = node.attributes || [];

            const attrs = Object.values(attributes).map((attr) =>
                attrToString(attr, options),
            );
            const printedAttrs =
                (attrs.length > 0 ? " " : "") + attrs.join(" ");

            if (
                node.children.length === 0 ||
                (!options.inlineErrors &&
                    node.children.every((c) => c.type === "error"))
            ) {
                return `<${nodeName}${printedAttrs} />`;
            }
            return `<${nodeName}${printedAttrs}>${content}</${nodeName}>`;
        }
        case "instruction": {
            const unsafe = /\?>/g;
            const subset = [">"];
            const nodeName = name(node.name);
            const result = escape(node.value, subset, unsafe);
            return "<?" + nodeName + (result ? " " + result : "") + "?>";
        }
        case "text": {
            let escapedText = escape(node.value, ["&", "<"]);
            if (options.doenetSyntax) {
                // A < symbols that are followed by whitespace is safe
                escapedText = escapedText.replace(/&lt;(?=\s|=)/g, "<");

                // We can replace &amp; with & except in the case where it would
                // accidentally make a character entity. For example `&amp;amp;` or `&amp;#x24;`
                escapedText = escapedText.replace(/&amp;(?!\S*;)/g, "&");
            }
            return escapedText;
        }
        case "error": {
            if (options.inlineErrors) {
                const errorElement: DastElement = {
                    type: "element",
                    name: "_error",
                    attributes: {},
                    children: [{ type: "text", value: node.message }],
                };
                return nodesToXml(errorElement, options);
            }
            return "";
        }
        case "macro": {
            if ((node as unknown as DastMacroV6).version === "0.6") {
                return macroToStringV6(node as any);
            }
            const macro = unwrappedMacroToString(node, options);

            let start = "$";
            let end = "";
            if (macroNeedsParens(node)) {
                start += "(";
                end += ")";
            }
            return start + macro + end;
        }
        case "function": {
            if ((node as unknown as DastMacroV6).version === "0.6") {
                return macroToStringV6(node as any);
            }
            const macro = unwrappedMacroToString(node, options);

            let start = "$$";
            let end = "";
            if (macroNeedsParens(node)) {
                start += "(";
                end += ")";
            }
            const args = node.input
                ? `(${node.input
                      .map((a) => nodesToXml(a, options))
                      .join(", ")})`
                : "";
            return start + macro + end + args;
        }
        default: {
            // Typescript exhaustiveness check
            const unusedType: void = node;
        }
    }

    if (!type) {
        throw new Error("Expected node, not `" + node + "`");
    }

    throw new Error("Cannot compile unknown node `" + type + "`");
}

/**
 * Serialize an attribute value.
 */
export function quote(value: string) {
    const result = String(value);
    let quoteMark = '"';

    const other = quoteMark === '"' ? "'" : '"';

    if (ccount(result, quoteMark) > ccount(result, other)) {
        quoteMark = other;
    }

    return quoteMark + escape(result, ["<", quoteMark]) + quoteMark;
}

/**
 * Convert a macro to a string, but do not wrap it in parens or add a `$` prefix.
 */
function unwrappedMacroToString(
    nodes: DastMacro | DastFunctionMacro,
    options: PrintOptions,
): string {
    const path = macroPathToString(nodes.path, options);
    let attrsStr = "";
    if (nodes.type === "macro") {
        const attrs = Object.values(nodes.attributes || {})
            .map((a) => attrToString(a, options))
            .join(" ");
        attrsStr = attrs.length > 0 ? `{${attrs}}` : "";
    }
    let propAccess = "";
    return path + attrsStr + propAccess;
}

function macroPathToString(
    path: DastMacroFullPath,
    options: PrintOptions,
): string {
    return path.map((part) => macroPathPartToString(part, options)).join(".");
}

function macroPathPartToString(
    pathPart: DastMacroPathPart,
    options: PrintOptions,
): string {
    return (
        pathPart.name +
        pathPart.index
            .map((part) => `[${nodesToXml(part.value, options)}]`)
            .join("")
    );
}

function attrToString(attr: DastAttribute, options: PrintOptions): string {
    if (attr.children.length === 0) {
        // Doenet syntax allows JSX style attributes without values assigned to them
        if (options.doenetSyntax) {
            return name(attr.name);
        }
        return `${name(attr.name)}="true"`;
    }
    return `${name(attr.name)}=${quote(nodesToXml(attr.children, options))}`;
}

function macroNeedsParens(macro: DastMacro | DastFunctionMacro): boolean {
    // We also might need wrapping if the path contains a `-` character
    return macro.path.some((part) => part.name.includes("-"));
}
