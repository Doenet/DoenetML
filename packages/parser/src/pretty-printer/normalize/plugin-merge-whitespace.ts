import { Plugin } from "unified";
import { DastRoot } from "../../types";
import { INDENTABLE_PRE_ELEMENTS, PRE_ELEMENTS } from "./special-nodes";
import { isElement, isText } from "./utils/testers";
import { SKIP, visit } from "./utils/visit";

/**
 * Unifiedjs plugin that converts sequences of whitespace to a single space character.
 */
export const mergeWhitespacePlugin: Plugin<void[], DastRoot, DastRoot> =
    function () {
        return (root: DastRoot) => {
            visit(root, (node, info) => {
                if (
                    isElement(node) &&
                    (PRE_ELEMENTS.has(node.name) ||
                        INDENTABLE_PRE_ELEMENTS.has(node.name))
                ) {
                    // Don't mangle whitespace inside of pre elements
                    return SKIP;
                }
                if (!isText(node)) {
                    // Only text nodes have whitespace to merge
                    return;
                }
                node.value = normalizeWhitespace(node.value);
            });
        };
    };

/**
 * Normalize whitespace in a string:
 *  - Lines that are only whitespace are are turned into single `\n` characters
 *  - Multiple whitespace characters in a row are turned into a single space character
 *  - Single newlines, but not double new lines, are turned into spaces
 */
export function normalizeWhitespace(str: string): string {
    // `[^\S\r\n]` matches all whitespace except for newlines

    // Make sure whitespace-only lines are turned into single newlines
    str = str.replace(/[^\S\r\n]*\n[^\S\r\n]*/g, "\n");

    let parts = str.split(/\n\n+/);
    parts = parts.map((part) => part.replace(/([^\S\r\n]|\n(?!\n))+/g, " "));

    // Make sure multiple whitespace characters in a row are turned into a single space
    // but do not replace two newlines in a row
    str = parts.join("\n\n");

    return str;
}
