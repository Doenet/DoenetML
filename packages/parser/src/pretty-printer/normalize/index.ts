import { Plugin, unified } from "unified";
import { DastRoot } from "../../types";
import { expandCdataPlugin } from "./plugin-expand-cdata";
import { mergeAdjacentTextPlugin } from "./plugin-merge-adjacent-text";
import { mergeWhitespacePlugin } from "./plugin-merge-whitespace";
import { removeEmptyTextNodesPlugin } from "./plugin-remove-empty-text-nodes";
import { removeInternodeWhitespacePlugin } from "./plugin-remove-internode-whitespace";
import { trimWhitespacePlugin } from "./plugin-trim-whitespace";

/**
 * Unifiedjs plugin that removes normalizes a DoenetML document. E.g.,
 * all whitespace is removed where it doesn't belong, cdata is expanded into text, etc..
 */
export const normalizeDoenetMlPlugin: Plugin<void[], DastRoot, DastRoot> =
    function () {
        const processor = unified()
            .use(expandCdataPlugin)
            .use(mergeAdjacentTextPlugin)
            .use(mergeWhitespacePlugin)
            .use(removeInternodeWhitespacePlugin)
            .use(trimWhitespacePlugin)
            .use(removeEmptyTextNodesPlugin);

        return (root: DastRoot, file) => {
            const processed = processor.runSync(root, file);
            return processed;
        };
    };

/**
 * Normalize a `DastRoot` AST.
 */
export function normalize(root: DastRoot) {
    const processor = unified().use(normalizeDoenetMlPlugin);
    return processor.runSync(root);
}
