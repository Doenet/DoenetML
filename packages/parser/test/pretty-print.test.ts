import { glob } from "glob";
import { describe, expect, it } from "vitest";
import * as fs from "node:fs/promises";
import { lezerToDast } from "../src/lezer-to-dast";
import { toXml } from "../src/dast-to-xml/dast-util-to-xml";
import { prettyPrint } from "../src/pretty-printer";
import util from "util";
import { normalizeWhitespace } from "../src/pretty-printer/normalize/plugin-merge-whitespace";

const origLog = console.log;
console.log = (...args) => {
    origLog(...args.map((x) => util.inspect(x, false, 10, true)));
};

describe("Prettier", async () => {
    it("Can normalize whitespace", () => {
        expect(normalizeWhitespace("a\nb")).toEqual("a b");
        expect(normalizeWhitespace("a \n\tb")).toEqual("a b");
        expect(normalizeWhitespace("\n  ")).toEqual(" ");
        expect(normalizeWhitespace("  \n  ")).toEqual(" ");
        expect(normalizeWhitespace("a\n\nb")).toEqual("a\n\nb");
        expect(normalizeWhitespace("a \n    \nb")).toEqual("a\n\nb");
        expect(normalizeWhitespace("a\n    \n   b")).toEqual("a\n\nb");
        expect(normalizeWhitespace("a\t\n    \n   b")).toEqual("a\n\nb");
        expect(normalizeWhitespace("a\t\n   \n\n \n   b")).toEqual("a\n\nb");
        expect(normalizeWhitespace("a  b")).toEqual("a b");
    });
    it("Space is trimmed from paragraphs", async () => {
        const cases = [
            { inStr: "<p> a </p>", outStr: "<p>a</p>" },
            { inStr: "<p>a</p>", outStr: "<p>a</p>" },
            { inStr: "<p>a\n</p>", outStr: "<p>a</p>" },
            { inStr: "<p>a\n \n</p>", outStr: "<p>a</p>" },
        ];
        for (const { inStr, outStr } of cases) {
            const prettyPrinted = await prettyPrint(inStr, {
                doenetSyntax: false,
                printWidth: 30,
            });
            expect(prettyPrinted).toEqual(outStr);
        }
    });
    it("Paragraphs show up on their own line", async () => {
        const cases = [
            { inStr: "<p>a</p><p>b</p>", outStr: "<p>a</p>\n<p>b</p>" },
        ];
        for (const { inStr, outStr } of cases) {
            const prettyPrinted = await prettyPrint(inStr, {
                doenetSyntax: false,
                printWidth: 30,
            });
            expect(prettyPrinted).toEqual(outStr);
        }
    });
    it("Double newlines are preserved but single newlines are turned into a space", async () => {
        const cases = [
            { inStr: "a\nb", outStr: "a b" },
            { inStr: "a\n  b", outStr: "a b" },
            { inStr: "a  \n  b", outStr: "a b" },
            { inStr: "a \n \n  b", outStr: "a\n\nb" },
        ];
        for (const { inStr, outStr } of cases) {
            const prettyPrinted = await prettyPrint(inStr, {
                doenetSyntax: false,
                printWidth: 30,
            });
            expect(prettyPrinted).toEqual(outStr);
        }
    });
    it("Prints function arguments", async () => {
        const cases = [
            { inStr: "$$f(x)", outStr: "$$f(x)" },
            { inStr: "$$f(x,y)", outStr: "$$f(x, y)" },
            { inStr: "$$f(x,$y)", outStr: "$$f(x, $y)" },
            { inStr: "$$f(x,$y z)", outStr: "$$f(x, $y z)" },
            { inStr: "$$f((),())", outStr: "$$f((), ())" },
            { inStr: "$$f(x,$$g(y,z))", outStr: "$$f(x, $$g(y, z))" },
            {
                inStr: "$$f(x,<math>alpha</math>)",
                outStr: "$$f(x, <math>alpha</math>)",
            },
            {
                inStr: "$$f((x),<math>alpha</math>)",
                outStr: "$$f((x), <math>alpha</math>)",
            },
            {
                inStr: "$$f(x,   <math>alpha</math>)",
                outStr: "$$f(x, <math>alpha</math>)",
            },
            {
                inStr: "<p>$$f(x, <math>alpha</math>)</p>",
                outStr: "<p>\n    $$f(x, <math>alpha</math>)\n</p>",
            },
        ];
        for (const { inStr, outStr } of cases) {
            const prettyPrinted = await prettyPrint(inStr, {
                doenetSyntax: false,
                printWidth: 30,
            });
            expect(prettyPrinted).toEqual(outStr);
        }
    });
    {
        // Regular parsing tests
        const files = glob
            .sync(
                new URL("./fragments/*pretty*.doenet", import.meta.url)
                    .pathname,
            )
            .filter((f) => !f.includes("-bad"));
        const fileMap: Record<string, string> = Object.fromEntries(
            files.map((f) => [f.split("/").pop(), f]),
        );

        for (const [filename, filepath] of Object.entries(fileMap)) {
            it(`${filename} pretty-prints correctly in Xml format and Doenet format`, async () => {
                const source = await fs.readFile(filepath, "utf-8");
                const prettyPrinted = await prettyPrint(source, {
                    doenetSyntax: false,
                    printWidth: 80,
                });
                //origLog(prettyPrinted);
                expect(prettyPrinted).toMatchSnapshot();
            });
        }
    }

    it("Always breaks children of BREAK_AROUND_ELEMENTS", async () => {
        const cases = [
            {
                inStr: "<graph><line /><line /></graph>",
                outStr: "<graph>\n    <line />\n    <line />\n</graph>",
            },
        ];
        for (const { inStr, outStr } of cases) {
            const prettyPrinted = await prettyPrint(inStr, {
                doenetSyntax: false,
                printWidth: 30,
            });
            expect(prettyPrinted).toEqual(outStr);
        }
    });
});
