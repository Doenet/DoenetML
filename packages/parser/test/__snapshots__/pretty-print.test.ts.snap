// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Prettier > pretty-print-1.doenet pretty-prints correctly in Xml format and Doenet format 1`] = `
"<text>a</text>
<p><textlist name="tl1" maxNumber="$mn1">a b c d e</textlist></p>
<p><textlist copysource="tl1" maxNumber="$mn2" name="tl2" /></p>

<p>Maximum number 2: <mathinput name="mn2" /></p>
<p>
    This is a really long sentence that is supposed to wrap eventually. Though,
    it is not supposed to rap musically...
</p>
<p>another p</p>

<p>
    Maximum number 1:
    <mathinput name="mn1" prefill="2" prop3="xxx" prop4="true" />
</p>

<choiceinput name="ci">
    <choice>yes</choice>
    <choice>no</choice>
    <choice>maybe</choice>
</choiceinput>

In document: $ci.choiceTexts <alert name="in_alert">$ci.choiceTexts</alert>
<blockquote name="in_blockquote">$ci.choiceTexts</blockquote>
<c name="in_c">$ci.choiceTexts</c><caption name="in_caption">
    $ci.choiceTexts
</caption><cell name="in_cell">$ci.choiceTexts</cell>
<choice name="in_choice">$ci.choiceTexts</choice>
<span name="in_span">$ci.choiceTexts</span><em name="in_em">$ci.choiceTexts</em>
<feedback name="in_feedback">$ci.choiceTexts</feedback>
<footnote name="in_footnote">$ci.choiceTexts</footnote>
<hint name="in_hint_w_title">
    <title>A title</title>
    $ci.choiceTexts
</hint>
<hint name="in_hint_wo_title">$ci.choiceTexts</hint>
<label name="in_label">$ci.choiceTexts</label>
<ul>
    <li name="in_li2">$ci.choiceTexts</li>
</ul>
<p name="in_p">$ci.choiceTexts</p>
<pre name="in_pre">$ci.choiceTexts
      x
        y
      </pre>
<p name="in_q"><q>$ci.choiceTexts</q></p>
<section name="in_section_w_title">
    <title name="sec_title">Title: $ci.choiceTexts</title>
    Text: $ci.choiceTexts
</section>
<section name="in_section_wo_title">
    $ci.choiceTexts
</section>
<solution name="in_solution">$ci.choiceTexts</solution>
<p name="in_sq"><sq>$ci.choiceTexts</sq></p>

<text name="in_text">$ci.choiceTexts</text>

<ol>
    <li name="in_li">$ci.choiceTexts</li>
    <li name="in_li">$ci.choiceTexts</li>

    <li name="in_li">$ci.choiceTexts</li>
</ol>"
`;
